structuredhub.go

```diff
@@ -0,0 +1,182 @@
+// their handler function.
+//
+// The data is serialized out using the marshaller and then back into  a
+// map[string]interface{}. If there is an error marshalling the data, Publish
+// fails with an error.  The resulting map is then updated with any
+// annotations provided. The annotated values are only set if the specified
+// field is missing or empty. After the annotations are set, the PostProcess
+// function is called if one was specified. The resulting map is then passed
+// to each of the subscribers.
+//
+// Subscribers are notified in parallel, and that no
+// modification should be done to the data or data races will occur.
+//
+// The channel return value is closed when all the subscribers have been
+// notified of the event.
+func (h *StructuredHub) Publish(topic Topic, data interface{}) (<-chan struct{}, error) {
+	if data == nil {
+		data = make(map[string]interface{})
+	}
+	asMap, err := h.toStringMap(data)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	for key, defaultValue := range h.annotations {
+		if value, exists := asMap[key]; !exists || value == reflect.Zero(reflect.TypeOf(value)).Interface() {
+			asMap[key] = defaultValue
+		}
+	}
+	if h.postProcess != nil {
+		asMap, err = h.postProcess(asMap)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+	}
+	h.hub.logger.Tracef("publish %q: %#v", topic, asMap)
+	return h.hub.Publish(topic, asMap), nil
+}
+
+func (h *StructuredHub) toStringMap(data interface{}) (map[string]interface{}, error) {
+	var result map[string]interface{}
+	resultType := reflect.TypeOf(result)
+	dataType := reflect.TypeOf(data)
+	if dataType.AssignableTo(resultType) {
+		cast, ok := data.(map[string]interface{})
+		if !ok {
+			return nil, errors.Errorf("%T assignable to map[string]interface{} but isn't one?", data)
+		}
+		return cast, nil
+	}
+	bytes, err := h.marshaller.Marshal(data)
+	if err != nil {
+		return nil, errors.Annotate(err, "marshalling")
+	}
+	err = h.marshaller.Unmarshal(bytes, &result)
+	if err != nil {
+		return nil, errors.Annotate(err, "unmarshalling")
+	}
+	return result, nil
+}
+
+// Subscribe takes a topic matcher, and a handler function. If the matcher
+// matches the published topic, the handler function is called with the
+// published Topic and the associated data.
+//
+// The handler function will be called with all maching published events until
+// the Unsubscribe method on the Unsubscriber is called.
+//
+// The hander function must have the signature:
+//   `func(Topic, map[string]interface{})`
+// or
+//   `func(Topic, SomeStruct, error)`
+// where `SomeStruct` is any structure. The map[string]interface{} from the
+// Publish call is unmarshalled into the `SomeStruct` structure. If there is
+// an error unmarshalling the handler is called with a zerod structure and an
+// error with the marshalling error.
+func (h *StructuredHub) Subscribe(matcher TopicMatcher, handler interface{}) (Unsubscriber, error) {
```

Comment:
I strongly object to this approach. This pushes the onus onto every subscriber to do the unmarshalling. The whole point of the structured hub is that it does the heavy lifting. We are told repeatedly that reflection is Go's way to deal with generics, so that is what I did. It keeps the reflection as a library implementation detail while offering the callers a clean interface. With the hub responsible for unmarshalling, we allow for future optimisations in caching the structures to pass to multiple subscribers if we feel the need. By pushing the onus onto the subscribers, you can't do this